MSC1960: OpenID information exchange with widgets (#1960)
matrix-org/matrix-doc: proposals/1960-integrations-openid.md

the request and as such need a way to get accurate information about who is contacting them.

This proposal introduces a way for widgets (room and account) to do so over the `fromWidget`
API proposed by [MSC1236](https://github.com/matrix-org/matrix-doc/issues/1236).

## Proposal

Room and account widgets may request new OpenID credentials from the user so they can log in/register with
the backing integration manager or other application. This is largely based on the prior art available
[here (element-web#7153)](https://github.com/vector-im/element-web/issues/7153). The rationale for such an
API is so that widgets can load things like a user's sticker packs or other information without having
to rely on secret strings. For example, a room could be used to let a user create custom sticker packs
via a common widget - it would be nice if that widget could auth the user without asking them to enter
their username and password into an iframe.

Widgets can request OpenID credentials from the user by sending a `fromWidget` action of `get_openid`
to initiate the token exchange process. The client responds with an acknowledgement of

> `{"state":"request"}` (or `{"state":"blocked"}` if the client/user doesn't think the widget is safe).

Reviewer: What is the value of `success` if `state` is `request`?

Author: https://github.com/matrix-org/matrix-doc/pull/1960#issuecomment-681979365

> `{"state":"request"}` (or `{"state":"blocked"}` if the client/user doesn't think the widget is safe).

Author: https://github.com/matrix-org/matrix-doc/pull/1960#issuecomment-681979365

The client then prompts the user if the widget should be allowed to get details about the user,
optionally providing a way for the user to always accept/deny the widget. If the user agrees, the
client sends a `toWidget` action of `openid_credentials` with `data` holding the raw OpenID credentials
object returned from the homeserver, and a `success: true` parameter, similar to the following:
```
{

>     "api": "fromWidget",

Reviewer: You've said above that this is `toWidget`?

Reviewer 2: I'm still not sure why this and the other two aren't `toWidget`?

Author: Because they are from the widget, not to the widget. This is a widget-initiated process.

Reviewer 2: Ahh, I see. So even though this is a response being sent from the client to the widget, we'd write `fromWidget` as the widget initiated the request in the first place.

Makes sense now, thank you.

Author: yea, the response format is (intentionally) a complete copy of the request with an added field

    "requestId": "AABBCC",
    "action": "openid_credentials",
    "widgetId": "DDEEFF",
    "data": {

>         "success": true,

Reviewer: In the `get_openid` request we have a `state` field that can be `blocked`, so it feels like rather than `success`, this should also have a `state` field which could be either `blocked` or `allowed`?

Author: https://github.com/matrix-org/matrix-doc/pull/1960#issuecomment-681979365

        "access_token": "SecretTokenHere",
        "token_type": "Bearer",
        "matrix_server_name": "example.com",
        "expires_in": 3600
    }
}
```

For clarity, the `data` consists of properties as returned by `/_matrix/client/r0/user/:userId/openid/request_token`
plus the `success` parameter.

If the user denies the widget, just `success: false` is returned in the `data` property.

To lessen the number of requests, a client can also respond to the original `get_openid` request with a
`state` of `"allowed"`, `success: true`, and the OpenID Connect credentials object (just like in the `data` for

> `openid_credentials`).

Reviewer 3: Is the following a valid example?

```json
{
    "api": "toWidget",
    "requestId": "AABBCC",
    "action": "get_openid",
    "widgetId": "DDEEFF",
    "data": {
        "state": "allowed",
        "success": true,
        "access_token": "SecretTokenHere",
        "token_type": "Bearer",
        "matrix_server_name": "example.com",
        "expires_in": 3600
    }
}

Author: https://github.com/matrix-org/matrix-doc/pull/1960#issuecomment-681979365

# MSC1960: OpenID Connect information exchange for widgets

Widgets are currently left with no options to verify the user's ID, making it hard for
personalized and authenticated widgets to exist. The spec says the `$matrix_user_id`

> template variable cannot be relied upon due to how easy it is to faslify, which is true.

Reviewer 2: ```suggestion
template variable cannot be relied upon due to how easy it is to falsify, which is true.
```

spec PR: https://github.com/matrix-org/matrix-doc/pull/2764

## Proposal

Typically widgets which need to accurately verify the user's identity will also have a
backend service of some kind. This backend service likely already uses the integration
manager authentication APIs introduced by [MSC1961](https://github.com/matrix-org/matrix-doc/pull/1961).

Through using the same concepts from MSC1961, the widget can verify the user's identity
by requesting a fresh OpenID Connect credential object to pass along to its backend, like
the integration manager which might be running it.

The protocol sequence defined here is based upon the previous discussion in the Element Web
issue tracker: https://github.com/vector-im/element-web/issues/7153

It is proposed that after the capabilities negotation, the widget can ask the client for
an OpenID Connect credential object so it can pass it along to its backend for validation.
The request SHOULD result in the user being prompted to confirm that the widget can have
their information. Because of this user interaction, it's not always possible for the user
to complete the approval within the 10 second suggested timeout by the widget spec. As

> such, the initial request by the widget can have one of three states:

Reviewer: It feels a bit out-of-order that we're talking about the responses before we've seen what the request is.

1. The client indicates that the user is being prompted (to be followed up on).
2. The client sends over credentials for the widget to verify.
3. The client indicates the request was blocked/denied.

The initial request from the widget looks as follows:

```json
{
    "api": "fromWidget",
    "action": "get_openid",
    "requestId": "AAABBB",
    "widgetId": "CCCDDD",
    "data": {}
}
```

Which then receives a response which has a `state` field alongside potentially the credentials

> to be verified. Matching the order of possible responses above, here are examples:

Reviewer: I would probably have just talked about what responses at the same time as the examples rather than having the reader refer back up to the list of possible responses, but ok - the content is fine, probably not worth re-wording.

```json
{
    "api": "fromWidget",
    "action": "get_openid",
    "requestId": "AAABBB",
    "widgetId": "CCCDDD",
    "data": {},
    "response": {
        "state": "request"
    }
}
```

```json
{
    "api": "fromWidget",
    "action": "get_openid",
    "requestId": "AAABBB",
    "widgetId": "CCCDDD",
    "data": {},
    "response": {
        "state": "allowed",
        "access_token": "s3cr3t",
        "token_type": "Bearer",
        "matrix_server_name": "example.org",
        "expires_in": 3600
    }
}
```

```json
{
    "api": "fromWidget",
    "action": "get_openid",
    "requestId": "AAABBB",
    "widgetId": "CCCDDD",
    "data": {},
    "response": {
        "state": "blocked"
    }
}
```

The credential information is directly copied from the `/_matrix/client/r0/user/:userId/openid/request_token`
response.

In the case of `state: "request"`, the user is being asked to approve the widget's attempt to
verify their identity. To ensure that future requests are quicker, clients are encouraged to
include a "remember this widget" option to make use of the immediate `state: "allowed"` or

> `state: "blocked"` responses above.

Reviewer 4: This seems to be redundant to what we have with the capability system. I would propose making this a capability as well. This would 
 - give the widget developer the choice to ask the user for oidc permissions on first startup.
 - the client developer does not need to track oidc permissions + capabilitis but can treat oidc as a capability without any custom logic.

There is no timeout associated with the user making their selection. Once a user does make
a selection (allow or deny the request), the client sends a `toWidget` request to indicate the
result, using a very similar structure to the above immediate responses:

```json
{
    "api": "toWidget",
    "action": "openid_credentials",
    "requestId": "EEEFFF",
    "widgetId": "CCCDDD",
    "data": {
        "state": "allowed",

>         "original_request_id": "AAABBB",

Reviewer: The mix of snake & camel case is unfortunate but this is probably the least terrible way of combining the two worlds.

Author: yea, it's a mess. It is however consistent with the other API actions (which are also awful).

[here (riot-web#7153)](https://github.com/vector-im/riot-web/issues/7153). The rationale for such an
API is so that widgets can load things like a user's sticker packs or other information without having
to rely on secret strings. For example, a room could be used to let a user create custom sticker packs
via a common widget - it would be nice if that widget could auth the user without asking them to enter
their username and password into an iframe.

Widgets can request OpenID credentials from the user by sending a `fromWidget` action of `get_openid`
to intiate the token exchange process. The client should respond with an acknowledgement of
`{"state":"request"}` (or `{"state":"blocked"}` if the client/user doesn't think the widget is safe).
The client should then prompt the user if the widget should be allowed to get details about the user,
optionally providing a way for the user to always accept/deny the widget. If the user agrees, the
client sends a `toWidget` action of `openid_credentials` with `data` holding the raw OpenID credentials
object returned from the homeserver, and a `success: true` parameter. If the user denies the widget,

> just `success: false` is returned in the `data` property. To lessen the number of requests, a client may

Reviewer: ok - sounds like `success` is mostly there to represent whether the request has been successful and all the info is actually in the `state` field?

> just `success: false` is returned in the `data` property. To lessen the number of requests, a client may

Author: pretty much, yes.

> also respond to the original `get_openid` request with a `state` of `"allowed"`, `success: true`, and

Reviewer: fair enough

## Proposal

> Room and account widgets may request new OpenID credentials from the user so they can log in/register with

Reviewer 2: ```suggestion
Room and account widgets may request new OpenID Connect credentials from the user so they can log in/register with
```

the backing integration manager or other application. This is largely based on the prior art available
[here (element-web#7153)](https://github.com/vector-im/element-web/issues/7153). The rationale for such an
API is so that widgets can load things like a user's sticker packs or other information without having
to rely on secret strings. For example, a room could be used to let a user create custom sticker packs
via a common widget - it would be nice if that widget could auth the user without asking them to enter
their username and password into an iframe.

> Widgets can request OpenID credentials from the user by sending a `fromWidget` action of `get_openid`

Reviewer 2: ```suggestion
Widgets can request OpenID Connect credentials from the user by sending a `fromWidget` action of `get_openid`
```

to initiate the token exchange process. The client responds with an acknowledgement of
`{"state":"request"}` (or `{"state":"blocked"}` if the client/user doesn't think the widget is safe).
The client then prompts the user if the widget should be allowed to get details about the user,
optionally providing a way for the user to always accept/deny the widget. If the user agrees, the

> client sends a `toWidget` action of `openid_credentials` with `data` holding the raw OpenID credentials

Reviewer 2: ```suggestion
client sends a `toWidget` action of `openid_credentials` with `data` holding the raw OpenID Connect credentials
```

object returned from the homeserver, and a `success: true` parameter, similar to the following:
```
{
    "api": "fromWidget",
    "requestId": "AABBCC",
    "action": "openid_credentials",
    "widgetId": "DDEEFF",
    "data": {
        "success": true,
        "access_token": "SecretTokenHere",
        "token_type": "Bearer",
        "matrix_server_name": "example.com",
        "expires_in": 3600
    }
}
```

For clarity, the `data` consists of properties as returned by `/_matrix/client/r0/user/:userId/openid/request_token`
plus the `success` parameter.

If the user denies the widget, just `success: false` is returned in the `data` property.

To lessen the number of requests, a client can also respond to the original `get_openid` request with a
`state` of `"allowed"`, `success: true`, and the OpenID Connect credentials object (just like in the `data` for
`openid_credentials`).

> The widget should not request OpenID credentials until after it has exchanged capabilities with the client,

Reviewer 2: ```suggestion
The widget should not request OpenID Connect credentials until after it has exchanged capabilities with the client,
```

> however this is not required to wait for the capabiltiies exchange.

Reviewer 2: ```suggestion
however it is not required to wait for the capabilities exchange to complete.
```

This is a little unclear to me. It sounds like a widget can send over its capabilities, before then immediately sending over creds without waiting for a response from the capabilities request?

Author: this is what it was before, which was confusing. I'll use different words per internal discussion.

> The widget acknowledges the `openid_credentials` request with an empty response object.

Reviewer 2: So that would be `{}` rather than:

```
{
    "api": "fromWidget",
    "requestId": "AABBCC",
    "action": "openid_credentials",
    "widgetId": "DDEEFF",
    "data": {}
}
```

right?

Author: https://github.com/matrix-org/matrix-doc/pull/1960#issuecomment-681979365
